### PR TITLE
[Merged by Bors] - Pre-allocate vectors in SSZ decoding

### PR DESCRIPTION
**Pull request by michaelsproul**
## Issue Addressed

Fixes a potential regression in memory fragmentation identified by @paulhauner here: https://github.com/sigp/lighthouse/pull/3371#discussion_r931770045.

## Proposed Changes

Immediately allocate a vector with sufficient size to hold all decoded elements in SSZ decoding. The `size_hint` is derived from the range iterator here:

https://github.com/sigp/lighthouse/blob/2983235650811437b44199f9c94e517e948a1e9b/consensus/ssz/src/decode/impls.rs#L489

## Additional Info

I'd like to test this out on some infra for a substantial duration to see if it affects total fragmentation.

### CODE REVIEW

**Comment by michaelsproul** [comment]
Running this for 24h on two similar Ropsten nodes didn't yield any significant change. If anything the vanilla v2.5.1 (light blue) code seems to be using slightly less free memory than this PR (dark blue)

total_memory

free_memory

I'm on the fence about whether we should bother merging it. On one hand explicit is better than implicit, so maybe we should be explicit about the allocation behaviour we want here rather than relying on the optimiser or the allocator to smooth it out for us (i.e. merge). On the other hand, if it ain't broke maybe we shouldn't meddle with it unnecessarily.

File: consensus/ssz/src/decode/try_from_iter.rs [hunk 1]
```diff
@@ -29,11 +29,18 @@ pub trait TryFromIter<T>: Sized {
 impl<T> TryFromIter<T> for Vec<T> {
     type Error = Infallible;
 
-    fn try_from_iter<I>(iter: I) -> Result<Self, Self::Error>
+    fn try_from_iter<I>(values: I) -> Result<Self, Self::Error>
     where
         I: IntoIterator<Item = T>,
     {
-        Ok(Self::from_iter(iter))
+        // Pre-allocate the max size of the Vec. We expect to receive vastly more valid
```
Comment (realbigsean) [inline consensus/ssz/src/decode/try_from_iter.rs]: I'm a little confused -- is this meant to pre-allocate the `max_size` of the variable list or `num_items` in the list? It looks like this allocates `num_items` based on the range you linked:

https://github.com/sigp/lighthouse/blob/2983235650811437b44199f9c94e517e948a1e9b/consensus/ssz/src/decode/impls.rs#L489

But the comments and variable names make me think the intent might be to allocate the max length of the variable list?

Regardless the change here does look equivalent to what we were doing prior to #3371

Comment (michaelsproul) [reply consensus/ssz/src/decode/try_from_iter.rs]: Oh yeah, allocating `num_items` is the intention. I've updated the comment to hopefully be a bit clearer about this

Comment (realbigsean) [reply consensus/ssz/src/decode/try_from_iter.rs]: ok this looks good to me then!

**Review by realbigsean — APPROVED** [review]

**Pull request closed without merging**